port/esp_idf/golioth_log_esp_idf.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2024 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <golioth/golioth_debug.h>
+#include <golioth/log.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include "string.h"
+
+static bool _cloud_log_enabled = CONFIG_GOLIOTH_AUTO_LOG_TO_CLOUD;
+
+static struct golioth_client *_client = NULL;
+
+static int golioth_log_espidf(const char *format,
+                          va_list args) {
+
+    const char *new_format = format;
+
+    if (!_cloud_log_enabled)
+    {
+        return -1;
+    }
+
+    if (!_client)
+    {
+        return -1;
+    }
+
+    // Avoid re-entering this function
+    static bool log_in_progress = false;
+    if (log_in_progress)
+    {
+        return -1;
+    }
+
+    va_list new_args;
+    va_copy(new_args, args);
+
+    // Since the format contains time (%lu) and tag (%s) values that need not
+    // be sent to the cloud like so : "[0;32mI (%lu) %s: Actual Log %d [0m",
+    // we remove the string upto colon : and update the args list
+    // Also, the 7th character in the string tells the log level as
+    // E = Error, W = Warning, I = info, D = Debug, V = Verbose
+    char log_level = format[7];
+    const char *colon_pos = strchr(format, ':');
+    if (colon_pos != NULL) {
+        new_format = colon_pos + 1; // Skip ':' and space
+
+        // Move va_list to correspond to the new format
+        va_arg(new_args, unsigned long); // Skip %lu
+        va_arg(new_args, const char *);   // Skip %s
+    }
+
+    // Figure out how large of a char buffer we need to store this message
+    int buffer_size = vsnprintf(NULL, 0, new_format, new_args) - 3;  // -3 to remove [0m at the end
```

Comment:
We can use a static buffer if we don't care much about the having [0m characters at the end of every log. This will be the case for logs shorter than the size of the static buffer. Longer log messages will get truncated. To get rid of the [0m characters at the end, we can find the string length and then add a null char at len - 3